app/services/add_attachment_service.rb

```diff
@@ -0,0 +1,63 @@
+#-- copyright
+# OpenProject is a project management system.
+# Copyright (C) 2012-2015 the OpenProject Foundation (OPF)
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See doc/COPYRIGHT.rdoc for more details.
+#++
+
+class AddAttachmentService
+  attr_reader :container, :author
+
+  def initialize(container, author:)
+    @container = container
+    @author = author
+  end
+
+  ##
+  # Adds and saves the uploaded file as attachment of the given container.
+  # In case the container supports it, a journal will be written.
+  #
+  # An ActiveRecord::RecordInvalid error is raised if any record can't be saved.
+  def add_attachment(uploaded_file:, description:)
+    attachment = Attachment.new(file: uploaded_file,
+                                container: container,
+                                description: description,
+                                author: author)
+    save attachment
```

Comment:
Don't worry, I am not offended. Just writing a lot ^^

Actually I didn't know of an existing service that provides me with an errors accessor. The ones I know usually involve calling two methods, which I wanted to avoid:

* `CreateWorkPackageService`
* `UpdateWorkPackageService`

What I liked about the exception (but this will also be solvable by the right accessors) is that it includes the faulty record and a client *could* infer further knowledge form the record itself, not only its errors (because sometimes `record` will be an attachment and sometimes a journable).

It would be great if you could point me to the other services you mentioned, usually this will make me accept that I went a wrong way xD